aws/resource_aws_datapipeline.go

```diff
@@ -0,0 +1,154 @@
+		Tags:     tagsFromMapDataPipeline(d.Get("tags").(map[string]interface{})),
+	}
+
+	if v, ok := d.GetOk("description"); ok && v != nil {
+		input.Description = aws.String(v.(string))
+	}
+
+	resp, err := conn.CreatePipeline(&input)
+
+	if err != nil {
+		return fmt.Errorf("Error creating datapipeline: %s", err)
+	}
+
+	d.SetId(aws.StringValue(resp.PipelineId))
+
+	return resourceAwsDataPipelineRead(d, meta)
+}
+
+func resourceAwsDataPipelineRead(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).datapipelineconn
+
+	v, err := resourceAwsDataPipelineRetrieve(d.Id(), conn)
+	if isAWSErr(err, datapipeline.ErrCodePipelineNotFoundException, "") || isAWSErr(err, datapipeline.ErrCodePipelineDeletedException, "") || v == nil {
+		log.Printf("[WARN] DataPipeline (%s) not found, removing from state", d.Id())
+		d.SetId("")
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Error describing DataPipeline (%s): %s", d.Id(), err)
+	}
+
+	d.Set("name", v.Name)
+	d.Set("description", v.Description)
+	d.Set("tags", tagsToMapDataPipeline(v.Tags))
+	return nil
+}
+
+func resourceAwsDataPipelineUpdate(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).datapipelineconn
+
+	if err := setTagsDataPipeline(conn, d); err != nil {
+		if isAWSErr(err, datapipeline.ErrCodePipelineNotFoundException, "") || isAWSErr(err, datapipeline.ErrCodePipelineDeletedException, "") {
+			log.Printf("[WARN] DataPipeline (%s) not found, removing from state", d.Id())
+			d.SetId("")
+			return nil
+		}
+
+		return fmt.Errorf("Error updating tags: %s", err)
+	}
+
+	return resourceAwsDataPipelineRead(d, meta)
+}
+
+func resourceAwsDataPipelineDelete(d *schema.ResourceData, meta interface{}) error {
+	conn := meta.(*AWSClient).datapipelineconn
+
+	opts := datapipeline.DeletePipelineInput{
+		PipelineId: aws.String(d.Id()),
+	}
+
+	_, err := conn.DeletePipeline(&opts)
+	if isAWSErr(err, datapipeline.ErrCodePipelineNotFoundException, "") || isAWSErr(err, datapipeline.ErrCodePipelineDeletedException, "") {
+		return nil
+	}
+	if err != nil {
+		return fmt.Errorf("Error deleting Data Pipeline %s: %s", d.Id(), err.Error())
+	}
+
+	return waitForDataPipelineDeletion(conn, d.Id())
+}
+
+func resourceAwsDataPipelineRetrieve(id string, conn *datapipeline.DataPipeline) (*datapipeline.PipelineDescription, error) {
+	opts := datapipeline.DescribePipelinesInput{
+		PipelineIds: []*string{aws.String(id)},
+	}
+
+	resp, err := conn.DescribePipelines(&opts)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.PipelineDescriptionList) != 1 ||
+		*resp.PipelineDescriptionList[0].PipelineId != id {
+		return nil, nil
+	}
+
+	return resp.PipelineDescriptionList[0], nil
+}
+
+func waitForDataPipelineDeletion(conn *datapipeline.DataPipeline, pipelineID string) error {
+	params := &datapipeline.DescribePipelinesInput{
+		PipelineIds: []*string{aws.String(pipelineID)},
+	}
+	return resource.Retry(10*time.Minute, func() *resource.RetryError {
```

Comment:
When using `resource.Retry()` we should always include a check against `isResourceTimeoutError()` in case the AWS Go SDK is being throttled and it never returns before the timeout. See also: #7873 e.g.

```go
err := resource.Retry(10*time.Minute, func() *resource.RetryError {
 /* ... existing logic here ... */
}

if isResourceTimeoutError(err) {
  _, err = conn.DescribePipelines(params) // the use of = instead of := is deliberate here
}

if isAWSErr(err, datapipeline.ErrCodePipelineNotFoundException, "") || isAWSErr(err, datapipeline.ErrCodePipelineDeletedException, "") {
  return nil
}

if err != nil {
  return fmt.Error("error reading DataPipeline Pipeline (%s) after deletion: %s", d.Id(), err)
}

return nil
```